### DISCUSSION TITLE
Recommendations for upgrade

### DISCUSSION THREAD

**Discussion by ACM-Unit** [OWASP/www-project-csrfguard#43] [Q&A] [answered]
Hi,
Do you have any recommendations for upgrade from 3.x to 4.x version?
We are using 3.1.0 version and would like to upgrade to 4.1.1 version

**Comment by forgedhallpass** [comment, marked as answer]
Hello @ACM-Unit,

Have you checked out the test application? It contains a default example setup.

**Reply by ACM-Unit** [reply]
Hello,
thanks for advice.
We've upgrade CsrfGuard to 4.1.1. version.
And now we've observed that we have no possibility to disable Token Per Page option, Looks like new functionality is strongly tied to this option. But mb we have some another way to disable it as it was before for our project?

**Reply by forgedhallpass** [reply]
Have you tried setting the `org.owasp.csrfguard.TokenPerPage` property to `false`?

**Reply by ACM-Unit** [reply]
yes, but in this case JavaScriptServlet send SC_BAD_REQUEST with message "This endpoint should not be invoked if the Token-Per-Page functionality is disabled!" 
JavaScriptServlet has checking 'if (csrfGuard.isTokenPerPageEnabled())'

**Reply by forgedhallpass** [reply]
I've just started the test application, with the `TokenPerPage` support disabled, and it works correctly for me. Maybe you still use the old JavaScript logic? The `JavaScriptServlet` has two endpoints. If `TokenPerPage` support is disabled, the solution relies on `doGet()` only, because the `doPost()` is for retrieving page tokens, hence the error message.

**Reply by ACM-Unit** [reply]
It works, thanks a lot
We updated script, but looks like after first app redeploy, those changes  weren't applied and I've turned on that flag
but now it works
